### PR TITLE
FOGL-1053 - unit tests for services/south/ingest.py

### PR DESCRIPTION
**Pull request by amarendra-dianomic**
Test result:
```
a@a ~/Development/FogLAMP (FOGL-1053) $ pytest -s -vv tests/unit/python/foglamp/services/south/test_ingest.py 
======================================================= test session starts ========================================================
platform linux -- Python 3.6.3, pytest-3.4.0, py-1.5.2, pluggy-0.6.0 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /home/a/Development/FogLAMP, inifile:
plugins: mock-1.6.0, cov-2.5.1, asyncio-0.8.0, allure-adaptor-1.7.10, aiohttp-0.3.0
collected 14 items                                                                                                                 

tests/unit/python/foglamp/services/south/test_ingest.py::TestIngest::test_read_config PASSED
tests/unit/python/foglamp/services/south/test_ingest.py::TestIngest::test_start PASSED
tests/unit/python/foglamp/services/south/test_ingest.py::TestIngest::test_stop PASSED
tests/unit/python/foglamp/services/south/test_ingest.py::TestIngest::test_increment_discarded_readings PASSED
tests/unit/python/foglamp/services/south/test_ingest.py::TestIngest::test__insert_readings SKIPPED
tests/unit/python/foglamp/services/south/test_ingest.py::TestIngest::test_write_statistics SKIPPED
tests/unit/python/foglamp/services/south/test_ingest.py::TestIngest::test_is_available_at_start PASSED
tests/unit/python/foglamp/services/south/test_ingest.py::TestIngest::test_is_available_at_stop PASSED
tests/unit/python/foglamp/services/south/test_ingest.py::TestIngest::test_is_available_when_all_lists_full PASSED
tests/unit/python/foglamp/services/south/test_ingest.py::TestIngest::test_add_readings_all_ok PASSED
tests/unit/python/foglamp/services/south/test_ingest.py::TestIngest::test_add_readings_if_stop PASSED
tests/unit/python/foglamp/services/south/test_ingest.py::TestIngest::test_add_readings_not_started PASSED
tests/unit/python/foglamp/services/south/test_ingest.py::TestIngest::test_add_readings_incorrect_data_values PASSED
tests/unit/python/foglamp/services/south/test_ingest.py::TestIngest::test_add_readings_when_one_list_becomes_full PASSED

=============================================== 12 passed, 2 skipped in 1.12 seconds ===============================================
```

Code coverage:

```
Coverage for python/foglamp/services/south/ingest.py : 57% Show keyboard shortcuts
279 statements   160 run 119 missing 0 excluded

```
Test Suite result:
```
a@a ~/Development/FogLAMP (FOGL-1053) $ pytest tests/unit/python/foglamp/
======================================================= test session starts ========================================================
platform linux -- Python 3.6.3, pytest-3.4.0, py-1.5.2, pluggy-0.6.0
rootdir: /home/a/Development/FogLAMP, inifile:
plugins: mock-1.6.0, cov-2.5.1, asyncio-0.8.0, allure-adaptor-1.7.10, aiohttp-0.3.0
collected 512 items                                                                                                                

tests/unit/python/foglamp/common/test_audit_logger.py ..........                                                             [  1%]
tests/unit/python/foglamp/common/test_configuration_manager.py ............................................................. [ 13%]
...........                                                                                                                  [ 16%]
tests/unit/python/foglamp/common/test_jqfilter.py ....                                                                       [ 16%]
tests/unit/python/foglamp/common/test_logger.py .....                                                                        [ 17%]
tests/unit/python/foglamp/common/test_process.py ........                                                                    [ 19%]
tests/unit/python/foglamp/common/test_service_record.py .......                                                              [ 20%]
tests/unit/python/foglamp/common/test_statistics.py ................                                                         [ 23%]
tests/unit/python/foglamp/common/microservice_management_client/test_microservice_management_client.py ..................... [ 27%]
..                                                                                                                           [ 28%]
tests/unit/python/foglamp/common/storage_client/test_payload_builder.py ...................................................s [ 38%]
...........                                                                                                                  [ 40%]
tests/unit/python/foglamp/common/storage_client/test_sc_exceptions.py ..........                                             [ 42%]
tests/unit/python/foglamp/common/storage_client/test_storage_client.py ...............                                       [ 45%]
tests/unit/python/foglamp/common/storage_client/test_utils.py ..................                                             [ 49%]
tests/unit/python/foglamp/common/web/test_middleware.py ......                                                               [ 50%]
tests/unit/python/foglamp/plugins/common/test_plugins_common_utils.py ....                                                   [ 50%]
tests/unit/python/foglamp/plugins/north/common/test_common.py .                                                              [ 51%]
tests/unit/python/foglamp/plugins/south/http_south/test_http_south.py ...............                                        [ 54%]
tests/unit/python/foglamp/services/common/test_avahi.py ......                                                               [ 55%]
tests/unit/python/foglamp/services/common/test_services_common_utils.py ....                                                 [ 56%]
tests/unit/python/foglamp/services/common/microservice_management/test_instance.py ..........                                [ 58%]
tests/unit/python/foglamp/services/core/test_connect.py ...                                                                  [ 58%]
tests/unit/python/foglamp/services/core/test_main.py .                                                                       [ 58%]
tests/unit/python/foglamp/services/core/api/test_audit.py .................                                                  [ 62%]
tests/unit/python/foglamp/services/core/api/test_backup_restore.py ...................................                       [ 68%]
tests/unit/python/foglamp/services/core/api/test_browser_assets.py ..........................................                [ 77%]
tests/unit/python/foglamp/services/core/api/test_certificate_store.py ..............                                         [ 79%]
tests/unit/python/foglamp/services/core/api/test_common_ping.py ...                                                          [ 80%]
tests/unit/python/foglamp/services/core/api/test_statistics_api.py ........                                                  [ 82%]
tests/unit/python/foglamp/services/core/interest_registry/test_interest_registry.py ........                                 [ 83%]
tests/unit/python/foglamp/services/core/scheduler/test_scheduler_entities.py .........                                       [ 85%]
tests/unit/python/foglamp/services/core/scheduler/test_scheduler_exceptions.py .....                                         [ 86%]
tests/unit/python/foglamp/services/core/service_registry/test_exceptions.py .....                                            [ 87%]
tests/unit/python/foglamp/services/core/service_registry/test_monitor.py ..                                                  [ 87%]
tests/unit/python/foglamp/services/core/service_registry/test_service_registry.py .........                                  [ 89%]
tests/unit/python/foglamp/services/south/test_ingest.py ....ss........                                                       [ 92%]
tests/unit/python/foglamp/services/south/test_services_south_main.py .                                                       [ 92%]
tests/unit/python/foglamp/services/south/test_services_south_server.py ...............                                       [ 95%]
tests/unit/python/foglamp/tasks/purge/test_purge.py ................                                                         [ 98%]
tests/unit/python/foglamp/tasks/purge/test_purge_main.py .                                                                   [ 98%]
tests/unit/python/foglamp/tasks/statistics/test_statistics_history.py ......                                                 [ 99%]
tests/unit/python/foglamp/tasks/statistics/test_statistics_main.py .                                                         [100%]

============================================== 509 passed, 3 skipped in 22.42 seconds ==============================================

```

### CODE REVIEW

**Comment by amarendra-dianomic** [comment]
@ashwinscale if you examine _insert_readings(), you will find nested infinite loops, so finding a way out will require more time and effort. Similarly, _write_statistics() will also need some tweaking to find a way out and hence this is also need more time and effort. Hence recommend to take up unit tests for these two methods in some future sprint.

File: tests/unit/python/foglamp/services/south/test_ingest.py [hunk 1]
```diff
@@ -0,0 +1,588 @@
...
+        "default": str(_max_readings_insert_batch_connection_idle_seconds)
+    },
+    "max_readings_insert_batch_reconnect_wait_seconds": {
+        "description": "The maximum number of seconds to wait before reconnecting to "
+                       "storage when inserting readings",
+        "type": "integer",
+        "default": str(_max_readings_insert_batch_reconnect_wait_seconds)
+    },
+}
+
```
Comment (praveen-garg) [inline tests/unit/python/foglamp/services/south/test_ingest.py]: @amarendra-dianomic not sure why do we need a copy paste of all above from original code file. should not we use these as imported var /def ?

File: tests/unit/python/foglamp/services/south/test_ingest.py [hunk 2]
```diff
@@ -0,0 +1,588 @@
...
+async def false_coro():
+    return True
+
+async def get_cat(old_config):
+    new_config = {}
+    for key, value in old_config.items():
+        new_value = copy.deepcopy(value)
+        new_value['value'] = new_value['default']
+        new_config.update({key: new_value})
+    return new_config
```
Comment (praveen-garg) [inline tests/unit/python/foglamp/services/south/test_ingest.py]: why? should not we get this as patched config manager?

File: tests/unit/python/foglamp/services/south/test_ingest.py [hunk 3]
```diff
@@ -0,0 +1,588 @@
...
+# FOGLAMP_END
+
+""" Test services/south/ingest.py
+
+"""
+import copy
+import pytest
+from unittest.mock import MagicMock
+from foglamp.services.south.ingest import *
+from foglamp.services.south import ingest
```
Comment (praveen-garg) [inline tests/unit/python/foglamp/services/south/test_ingest.py]: re-think about these 2 imports

File: tests/unit/python/foglamp/services/south/test_ingest.py (hunk 1 above)
Comment (amarendra-dianomic) [reply tests/unit/python/foglamp/services/south/test_ingest.py]: done via another way.

File: tests/unit/python/foglamp/services/south/test_ingest.py (hunk 2 above)
Comment (amarendra-dianomic) [reply tests/unit/python/foglamp/services/south/test_ingest.py]: this is better

File: tests/unit/python/foglamp/services/south/test_ingest.py (hunk 3 above)
Comment (amarendra-dianomic) [reply tests/unit/python/foglamp/services/south/test_ingest.py]: required

**Comment by amarendra-dianomic** [comment]
@praveen-garg feedback changes done. Pl check.

File: tests/unit/python/foglamp/services/south/test_ingest.py (hunk 3 above)
Comment (praveen-garg) [reply tests/unit/python/foglamp/services/south/test_ingest.py]: that should be available as `__LOGGER` with import * (no?)

File: tests/unit/python/foglamp/services/south/test_ingest.py [hunk 4]
```diff
@@ -0,0 +1,517 @@
...
+        # THEN
+        assert 1 == Ingest._discarded_readings_stats
+
+    @pytest.mark.asyncio
+    async def test__insert_readings(self, mocker):
+        # TODO: This method uses a while True loop. Investigate as to how to write unit test for an infinite loop.
+        pass
+
+    @pytest.mark.asyncio
+    async def _write_statistics(self, mocker):
```
Comment (praveen-garg) [inline tests/unit/python/foglamp/services/south/test_ingest.py]: this should be skipped with reason! and name `test__write_statistics`; 
its call is not covered in exposed def?

File: tests/unit/python/foglamp/services/south/test_ingest.py [hunk 5]
```diff
@@ -0,0 +1,517 @@
...
+    async def test_increment_discarded_readings(self, mocker):
+        # GIVEN
+        # WHEN
+        Ingest.increment_discarded_readings()
+
+        # THEN
+        assert 1 == Ingest._discarded_readings_stats
+
+    @pytest.mark.asyncio
+    async def test__insert_readings(self, mocker):
```
Comment (praveen-garg) [inline tests/unit/python/foglamp/services/south/test_ingest.py]: this should be skipped with reason!

File: tests/unit/python/foglamp/services/south/test_ingest.py (hunk 3 above)
Comment (amarendra-dianomic) [reply tests/unit/python/foglamp/services/south/test_ingest.py]: Unfortunately, not. It is not available with import *, yet to find a reason for it.

**Comment by amarendra-dianomic** [comment]
@praveen-garg feedback changes done. Pl check.

**Pull request merged**